Stepper optimisations

tombrazier opened the pull request:
### Description

A tranche of speed improvements for the stepper ISR along with an experimental new dynamic multi-stepping algorithm.

Needs testing and the dynamic muti-stepping algorithm needs to be made optional and extended to cover adaptive step smoothing.

More details to follow.

### Requirements

<!-- Does this PR require a specific board, LCD, etc.? -->

### Benefits

<!-- What does this PR fix or improve? -->

### Configurations

<!-- Attach Configurations ZIP and any other files needed to test this PR. -->

### Related Issues

#25117

tombrazier committed:
For the fastest step rates, do the least work in calc_timer_interval()

tombrazier committed:
When moving fast, do not waste CPU cycles applying an offset to improve the minimal speed case

tombrazier committed:
Save 256 bytes of PROGMEM and gain cycles by saving the LERP gradient as a byte for speed_lookuptable_fast

tombrazier committed:
Use hal_timer_t for timers in the stepper ISR

tombrazier committed:
The compiler is not smart enough to distribute the multiplication through the ternay function

tombrazier committed:
Removed redundant check against step_needed

tombrazier committed:
Use the AxisFlags bitfield rather than the xyze_bool_t array for step_needed

tombrazier committed:
Add 1 byte to RAM and save a lot of load/compare of 32 bit numbers

tombrazier committed:
Helped the compiler to keep bresenham logic in registers

tombrazier committed:
As well as not needing to disable interrupts on AVR, you do not need to re-enable them either

tombrazier committed:
Added multistepping debugging to max7219

tombrazier committed:
Not clean: added dynamic CPU control

tombrazier committed:
Merge branch 'bugfix-2.1.x' into stepper_optimisations

thinkyhead committed:
remove dead friends

thinkyhead committed:
Update createSpeedLookupTable.py

thinkyhead committed:
sei() automatically invoked by AVR reti

thinkyhead committed:
Minor calc_timer_interval optimizations

thinkyhead committed:
style tweak

thinkyhead committed:
style tweak

tombrazier committed:
Corrected logic for speed_lookuptable_fast to match values from 27cca14, matched formatting to speed_lookuptable.h and improved rounding

tombrazier commented:
@thinkyhead In commit 27cca14 the actual values for `speed_lookuptable_fast` changed. The CPU*2 offset was removed. I also added in rounding for `speed_lookuptable_fast` but, oddly, not for `speed_lookuptable_slow`. Latest commit cleans all this up, matching the `speed_lookuptable_fast` I previously committed and adding rounding to `speed_lookuptable_slow`. Also cleaned up formatting to match the actual header file as we have it.

tombrazier committed:
Optimisation of dynamic multi-steppping

tombrazier committed:
Don't update max7219 from within stepper ISR

tombrazier commented:
Oh man, this has been is doing my head in. I have spent ages trying to figure out why performance is better when the `loops` parameter to `Stepper::calc_timer_interval(uint32_t step_rate, uint8_t loops)` is passed by reference rather than by value. And, bizarrely, it's even better in completely unrelated functions!

Using `-S` to generate assembly confirms that there is almost no change to the assembler code except for some extra instructions in the pass by reference cases which should make performance slightly worse!

I finally noticed that binary size is bigger in the pass by value version. So I now think it's the link time optimiser inlining the pass by value version of the function and thereby both increasing code size and decreasing performance.

thinkyhead commented:
Ok, so checking the old code, we have 4 calls to `calc_timer_interval(a, b)` and as you point out, `steps_per_isr` is being passed by reference in all cases.
```
 2160: interval      = calc_timer_interval(acc_step_rate << oversampling_factor, steps_per_isr);
 2230: interval      = calc_timer_interval(step_rate << oversampling_factor, steps_per_isr);
 2292: ticks_nominal = calc_timer_interval(current_block->nominal_rate << oversampling_factor, steps_per_isr);
 2611: interval      = calc_timer_interval(current_block->initial_rate << oversampling_factor, steps_per_isr);
```
An important part of this `calc_timer_interval` method is that it modifies `loops`, thus `steps_per_isr` with the line:
```cpp
loops = multistep;
```
Your updated code removes the pass-by reference for `loops` so that `steps_per_isr` is not being modified by the call. That seems like an important difference. This probably has bearing on the way that parameters are being passed, as your analysis found.
So, you are correct: We could modify the name of the method (e.g., `calc_timer_interval_and_steps`) and drop the `loops` parameter and have it modify `steps_per_isr` directly. So I'll submit a commit that makes that change, then we can compare and see if that ends up preserving the more optimal code for the benefit of AVR.

thinkyhead committed:
Use steps_per_isr directly

thinkyhead commented:
I see that `MAX_STEP_ISR_FREQUENCY_1X` is the only macro of its type now in use with your changes. Thus we can drop the `R` parameter and substitute 0 throughout. This results in:
```cpp
#define ISR_LOOP_CYCLES _MAX(MIN_ISR_LOOP_CYCLES, MIN_STEPPER_PULSE_CYCLES)
#define ISR_SHAPING_LOOP_CYCLES TERN0(HAS_SHAPING, (ISR_LOOP_BASE_CYCLES) + TERN0(INPUT_SHAPING_X, ISR_X_STEPPER_CYCLES) + TERN0(INPUT_SHAPING_Y, ISR_Y_STEPPER_CYCLES))
#define ISR_EXECUTION_CYCLES() (ISR_BASE_CYCLES + ISR_S_CURVE_CYCLES + ISR_SHAPING_BASE_CYCLES + ISR_LOOP_CYCLES + ISR_SHAPING_LOOP_CYCLES + ISR_LA_BASE_CYCLES + ISR_LA_LOOP_CYCLES)
#define MAX_STEP_ISR_FREQUENCY_1X ((F_CPU) / ISR_EXECUTION_CYCLES())
```

thinkyhead committed:
Actually, don't modify

thinkyhead committed:
Drop MAX_STEP_ISR_FREQUENCY_*X

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25474

thinkyhead committed:
cosmetic

thinkyhead commented:
So, I've done a thorough review of your changes, except for the `time_spent` portion, and everything in that slice is looking very sensible and justified. I pushed a branch with the "most obvious" parts which you can see here: https://github.com/MarlinFirmware/Marlin/compare/bugfix-2.1.x...thinkyhead:Marlin:bf2_base_stepper_opts_PR

The new `time_spent` logic which moves the handling of `steps_per_isr` is the only "tricky" bit that I need to look at more closely to fully grok.

I see that in `calc_timer_interval` you put the `step_rate <= min_step_rate` condition first. Is that because this is the most likely case to evaluate true most of the time? And also, I see that `step_rate -= min_step_rate` happens only after the test for `step_rate >= 0x0800` instead of before. Is that due to changes in the speed lookup table, a fix for bad logic, or…? I'll have to look back in history to see why we did things in that particular way.

It occurs to me that we could save 24 whole bytes of flash (!) by using `&speed_lookuptable_fast[uint8_t(step_rate >> 8) - 8]` since the compiler should treat it something like `(&speed_lookuptable_fast - 24)[step_rate >> 8]`, but it's a toss-up whether to do that, since 24 bytes isn't a huge amount of savings.

Anyway, if we want to reduce this PR down to the `time_spent` logic for additional review and commit the "obvious" portion, I can go ahead and do that at any time. I'll wait until after fixing up `calc_timer_interval(rate)` to match this PR before doing that, assuming the `step_rate` logic is sound.

thinkyhead commented:
Reaching back into the annals of time, I see that I made this change in May 2018 with 6919e87656…
```diff
- NOLESS(step_rate, uint32_t(F_CPU / 500000U));
- step_rate -= F_CPU / 500000; // Correct for minimal speed
+ constexpr uint32_t min_step_rate = F_CPU / 500000U;
+ NOLESS(step_rate, min_step_rate);
+ step_rate -= min_step_rate; // Correct for minimal speed
```
…so originally it was meant to first keep the step rate at or above minimum, and then to subtract `min_step_rate` to keep the lookup zero-based, I presume.

Prior to that the `step_rate` was "flattened out" in Aug 2017 with 4b16fa3272 by @p3p and me, ahead of the test `step_rate >= 8 * 256`.

I guess the idea there was to treat any step rate below the minimum as "instantaneous", and I believe the comparison to 0x800 has to do with the size of the lookup table being 0x400 for the benefit of `table_address += ((step_rate) >> 1) & 0xFFFC`.

So, it's somewhat inconclusive whether the subtraction of `min_step_rate` should come before that comparison. But my best understanding is that it has to do with the lookup table, and so I lean towards the conclusion that the subtraction should indeed come before the comparison.

tombrazier commented:
> So I'll submit a commit that makes that change, then we can compare and see if that ends up preserving the more optimal code for the benefit of AVR.

Eliminating the `loop` parameter entirely still results in a less efficient use of the stack frame in `Stepper::isr()`! However that is with my timing test code also taking up local variable space and before any other optimisations to come (of which I have at least one in mind). I guess a final test once all the changes are in place might be needed.

I see you spotted the error in f5b69e4. With my changes `steps_per_isr` should not be updated from `calc_timer_interval_and_steps()`. The `loops` variable only gets shifted down to 1 as a way of counting how many bits to shift `step_rate`.

> I see that MAX_STEP_ISR_FREQUENCY_1X is the only macro of its type now in use with your changes.

I am still thinking of putting the old multi-stepping logic back and making it selectable with a config macro. Just in case. OTOH that would result in messier code. What do you think? If we don't put it back I think we need a lot of testing.

I am also thinking of extending the dynamic multi-stepping logic to determine oversampling dynamically. (Which will be really nice - we'd get oversampling in the cheap cruise phase even when the more expensive ramps require no oversampling.) This would remove the another usage of `MAX_STEP_ISR_FREQUENCY_1X`. However it would also mean
https://github.com/MarlinFirmware/Marlin/blob/3e12934ea0878c311b55c00766493f1b7cda0044/Marlin/src/module/stepper.h#L337-L344
could no longer give a potential SRAM saving on AVRs with ASS and IS.

The only remaining use should also be removed, I think:
https://github.com/MarlinFirmware/Marlin/blob/3e12934ea0878c311b55c00766493f1b7cda0044/Marlin/src/module/stepper.cpp#L2101
This line is buggy. It changes `step_rate` and thereby breaks assumptions that `Stepper::block_phase_isr()` makes, causing weird behaviour which @MusicJelly describes [here](https://github.com/MarlinFirmware/Marlin/issues/25117#issuecomment-1414779658). I think the `DISABLE_MULTI_STEPPING` case should just be brought into line with the multi-stepping case where no speed limit is applied and the CPU just maxes out if you try to drive axes too fast.

> I pushed a branch with the "most obvious" parts which you can see here

It's a bit confusing trying to compare your branch with this PR. I think the only changes are:
- the code in `calc_timer_interval` is a bit different, particularly the changes in 52771c5c7ccccf9be3e23b97f5ab5bfd2dafa96a are missing.
- it doesn't include all the `time_spent` logic.

> I see that in calc_timer_interval you put the step_rate <= min_step_rate condition first. Is that because this is the most likely case to evaluate true most of the time? And also, I see that step_rate -= min_step_rate happens only after the test for step_rate >= 0x0800 instead of before. Is that due to changes in the speed lookup table, a fix for bad logic, or…? I'll have to look back in history to see why we did things in that particular way.

I put the `step_rate >= 0x0800` test first because that is the scenario where the step rate is highest and where we need the function to complete as quickly as possible.

`step_rate -= min_step_rate` is needed when the step rate is slow. In addition to your analysis above, it avoids having to store the result of a divide by zero in `speed_lookuptable_slow[0]`. The zeroeth element of `speed_lookuptable_fast` is never used so there is no need to deal with divide by zero in that case. So I removed the subtraction to save CPU cycles and adjusted `speed_lookuptable_fast` accordingly.

Both of the latter two changes are aimed at making `calc_timer_interval()` as fast a possible for high step rates, which is when we really need it to perform.

> It occurs to me that we could save 24 whole bytes of flash (!) by using &speed_lookuptable_fast[uint8_t(step_rate >> 8) - 8]

It occurred to me too. The compiler did not see things the same way and added extra logic for the `- 8`. Witness:
`const uintptr_t table_address = uintptr_t(&speed_lookuptable_fast[uint8_t(step_rate >> 8)]);`
generates
```
	ldi r18,lo8(3)
	mul r23,r18
	movw r20,r0
	clr __zero_reg__
	subi r20,lo8(-(_ZL22speed_lookuptable_fast))
	sbci r21,hi8(-(_ZL22speed_lookuptable_fast))
```
whereas
`const uintptr_t table_address = uintptr_t(&speed_lookuptable_fast[uint8_t(step_rate >> 8) -8]);`
generates
```
	mov r20,r23
	subi r20,8
	sbc r21,r21
	movw r30,r20
	movw r18,r20
	lsl r18
	rol r19
	add r18,r30
	adc r19,r31
	subi r18,lo8(-(_ZL22speed_lookuptable_fast))
	sbci r19,hi8(-(_ZL22speed_lookuptable_fast))
```

On the other hand
`const uintptr_t table_address = uintptr_t(&(&speed_lookuptable_fast[uint8_t(step_rate >> 8)])[-8]);`
```
	ldi r18,lo8(3)
	mul r23,r18
	movw r20,r0
	clr __zero_reg__
	subi r20,lo8(-(_ZL22speed_lookuptable_fast-24))
	sbci r21,hi8(-(_ZL22speed_lookuptable_fast-24))
```
but it's pretty confusing code!

## Other things to do

- There are a few more things I want to test.
- I have an optimisation in mind for `count_position[_AXIS(AXIS)] += count_direction[_AXIS(AXIS)]`. The avr compiler generated code takes 24 cycles per axis for this. I think I can reduce that to an average of 9.5 cycles per axis.
- I think we should add MULTI_STEPPING_LIMIT to the config so a user can specify the micro-stepping level of their stepper with the lowest micro-stepping setting. That way if we exceed the MCU's capabilities it just results in axes not going any faster and the LCD becoming unresponsive rather than lost steps.

> Anyway, if we want to reduce this PR down to the time_spent logic for additional review and commit the "obvious" portion, I can go ahead and do that at any time

I think this makes sense once:
- The `calc_timer_interval()` code is resolved.
- I have done the other few things I mentioned above.
- I have worked out whether passing `loops` by reference still gives the best performance.

tombrazier committed:
Added MULTISTEPPING_LIMIT to config to limit the effect of multistepping. This replaces DISABLE_MULTI_STEPPING which was never visible in the config files. And removed the speed cap from calc_timer_interval() for DISABLE_MULTI_STEPPING because it breaks the calculations in block_phase_isr()

thinkyhead committed:
tweaks to allow CI test

thinkyhead committed:
MULTISTEPPING_LIMIT must be simple

thinkyhead committed:
comment tweak

thinkyhead committed:
comment tweak

stepper driver "micro-steps" has no direct bearing on step loss per se

thinkyhead committed:
move conditional

thinkyhead committed:
reorder calc_timer_interval for comparison

thinkyhead committed:
"calc_multistep_timer_interval"

thinkyhead committed:
filter by multistepping limit

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25474

thinkyhead committed:
general cleanups from review

tombrazier commented:
I quite liked the commented code in `if (loops >=  2) { step_rate >>= 1; /*loops >>= 1; */ }`. Makes it clear that the shift is logically right but unneeded.

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -1522,7 +1523,7 @@ void Stepper::isr() {
     #endif
 
     // Get the interval to the next ISR call
-    const uint32_t interval = _MIN(
```
good spotting

tombrazier committed:
Signficant optimisation of input shaping

thinkyhead committed:
format tweak

thinkyhead committed:
branch-not-taken is faster…

…but is this the more common case?

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25474

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25474

thinkyhead committed:
remove Q

thinkyhead commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2091,45 +2113,11 @@ hal_timer_t Stepper::calc_timer_interval(uint32_t step_rate) {
+    uint8_t loops = steps_per_isr;
+    if (MULTISTEPPING_LIMIT >= 16 && loops >= 16) { step_rate >>= 4; loops >>= 4; }
+    if (MULTISTEPPING_LIMIT >=  4 && loops >=  4) { step_rate >>= 2; loops >>= 2; }
+    if (MULTISTEPPING_LIMIT >=  2 && loops >=  2) { step_rate >>= 1; }
```
Quick question: Is this approach as accurate as the `limit` table? I have a vague suspicion that the `(F_CPU) / ISR_EXECUTION_CYCLES(n)` values might have one or more extra digits that carry, and those might be getting lost if we are just shifting them out. But, I assume you already checked.

tombrazier replied:
From the perspective of `step_rate`, which this code is calculating, there is no change. It was right shifted by a certain number of bits in the old code and that's still the case in the new code. What is different is how we derive the number of bits to shift by.

thinkyhead commented on Marlin/src/module/stepper.cpp:
```diff
@@ -194,6 +194,8 @@ uint32_t Stepper::acceleration_time, Stepper::deceleration_time;
   uint8_t Stepper::steps_per_isr = 1; // Count of steps to perform per Stepper ISR call
 #endif
 
+hal_timer_t Stepper::time_spent_in_isr = 0, Stepper::time_spent_out_isr = 0;
```
I'm thinking of making a (hidden) option to turn this off and use the existing timing method, in the unlikely event that there is some edge case we haven't seen yet. One of the things that occurred to me was a potential situation where we do the timing calculation for a set of small identical segments, and as a result of reducing the multi-stepping based on the timing of one segment, the `time_spent_in_isr` goes down just enough to make the next segment calculation favor more multi-steps, which makes `time_spent_in_isr` go up just enough to make the next segment calculation favor fewer multi-steps, which makes `time_spent_in_isr` go down just enough to make the next segment calculation favor more multi-steps….

Is that something that could actually occur with just the right combination of segment-length and feedrate?

tombrazier replied:
I agree with the option to select the old behaviour. As you say, who knows what edge case there might be out there.

On the specific question about the behviour continually flipping between more and fewer multi-steps: If flipping like this happened, it could occur every time `block_phase_isr()` runs. However when multi-stepping doubles the amount of time required by the ISR does not decrease by as much as half. And since the ISR needs to consume less than 50% of the processor time before multi-stepping will halve again, I am pretty sure there is sufficient hysteresis to prevent continual flipping.

thinkyhead committed:
Old adaptive multi-stepping option

thinkyhead committed:
tweak e motor direction

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/25474

tombrazier commented:
Awesome.
